source/common/common/callback_impl.h

```diff
@@ -81,5 +84,107 @@ template <typename... CallbackArgs> class CallbackManager {
   const std::shared_ptr<bool> still_alive_{std::make_shared<bool>(true)};
 };
 
+/**
+ * Utility class for managing callbacks. Thread-safe.
+ * @see CallbackManager
+ */
+template <typename... CallbackArgs> class ThreadSafeCallbackManager {
+public:
+  using Callback = std::function<void(CallbackArgs...)>;
+
+  /**
+   * Add a callback.
+   * @param callback supplies the callback to add.
+   * @return CallbackHandlePtr a handle that can be used to remove the callback.
+   */
+  ABSL_MUST_USE_RESULT CallbackHandlePtr add(Callback callback) {
+    Thread::LockGuard lock(lock_);
+    auto new_callback = std::make_unique<CallbackHolder>(*this, callback);
+    callbacks_.emplace_back(new_callback.get());
+    // Get the list iterator of added callback handle, which will be used to remove itself from
+    // callbacks_ list.
+    new_callback->it_ = (--callbacks_.end());
+    return new_callback;
+  }
+
+  /**
+   * Run all registered callbacks.
+   * NOTE: This code is currently safe if a callback deletes ITSELF from within a callback. It is
+   *       not safe if a callback deletes other callbacks. If that is required the code will need
+   *       to change (specifically, it will crash if the next callback in the list is deleted).
```

Comment:
Are there tests that cover this removal while iterating behavior?

I think that `remove` attempts to take lock_ despite remaining held while the callback executes.  IIRC the locks used here are not re-entrant, attempts to take a lock that is already held will deadlock.

Reply:
- [ ] Add test for removal while iterating.

Reply:
You are correct that the lock is not re-entrant and rather than trying to determine a clever way of not needing a lock, I've simply documented that removal is not a supported operation _within_ the callback. Let me know if this is an acceptable trade-off.

Reply:
No longer true after refactor of callback manager.